bundles/org.openhab.core.audio/src/main/java/org/openhab/core/audio/AudioSinkAsync.java

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright (c) 2010-2023 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.audio;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.common.Disposable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Definition of an audio output like headphones, a speaker or for writing to
+ * a file / clip.
+ * Helper class for asynchronous sink : when the process() method returns, the {@link AudioStream}
+ * may or may not be played. It is the responsibility of the implementing AudioSink class to
+ * complete the CompletableFuture when playing is done. Any delayed tasks will then be performed, such as volume
+ * restoration.
+ *
+ * @author Gwendal Roulleau - Initial contribution
+ */
+@NonNullByDefault
+public abstract class AudioSinkAsync implements AudioSink {
+
+    private final Logger logger = LoggerFactory.getLogger(AudioSinkAsync.class);
+
+    private final Map<AudioStream, CompletableFuture<@Nullable Void>> runnableByAudioStream = new HashMap<>();
+
+    @Override
+    public CompletableFuture<@Nullable Void> processAndComplete(@Nullable AudioStream audioStream)
+            throws UnsupportedAudioFormatException, UnsupportedAudioStreamException {
+        CompletableFuture<@Nullable Void> completableFuture = new CompletableFuture<@Nullable Void>();
+        try {
+            if (audioStream != null) {
+                runnableByAudioStream.put(audioStream, completableFuture);
+            }
+            processAsynchronously(audioStream);
+            return completableFuture;
+        } finally {
+            if (audioStream == null) {
```

Comment:
If this is true, no `CompletableFuture` was added, so it doesn't need to be removed. In fact, it would be better to do the null check above and return a `CompletableFuture.completedFuture(null)`. You can skip the try/catch block then.

Reply:
Indeed, the remove in the null case is an error.

But, correct me if I'm wrong, as we have to call the processAsynchronously method -even if the stream is null, in order to ask the sink to stop current playing-, then if I remove the try / finally, I cannot be sure that the delayed volume restauration will occur (if there is an exception during the processAsynchronously(null) method).